### PR TITLE
Moddable Milking Monsters

### PR DESCRIPTION
**Pull request by AlecWhite**
I wasn't too happy with the previous https://github.com/CleverRaven/Cataclysm-DDA/pull/21272 way it was implemented as it didn't left much room for moddability and it wasn't as fun as this one.
So I re-remade it using the previous one as base and tweaking here and there so I don't have to force the inclusion of new variables for monster class and type.

-Added monexamine.h/cpp to take care of milking interaction and any other interactions with pets or monsters that we make in the future.
-Added MILKABLE flag and milked effect.
-Added a PET_WONT_FOLLOW flag, monsters with this flag when tamed will not follow the player.
-Cows are now milkable and gives milk item 4 times per 24 hours, lowered aggression and added the PET_WONT_FOLLOW flag.
-Refactored the pet_food code to use the PetFood enum, and changed all the if to switches, this way is easier to add future items to tame other animals.
-Added cattle food recipe, item and action.
-You need first to tame a cow with it's cattle food before milking it.
-Milked items of milkable monster depends on starting_ammo, where ammo type is the item given and ammo count the times per 24 hours you can milk it. So that milkable animals can be further modded to your heart's content, and avoids the inclusion of more variables.
-Added check so that only liquid phase items can be milked.
-All monsters with milkable flag can be tamed with cattle feed.

Once more, thank you to everyone that helps me reviewing the code.

I hope the third time's a charm and I'm quite happy with how this version lets people mod it by just editing the JSONs.

### CODE REVIEW

File: data/json/items/tools.json [hunk 1]
```diff
@@ -8270,5 +8270,20 @@
     "ammo": "battery",
     "initial_charges": 100,
     "max_charges": 100
+  },
+  {
+    "id": "cattlefodder",
+    "type": "TOOL",
+    "name": "cattle fodder",
+    "name_plural": "cattle fodder",
+    "description": "What cattle eats. Mainly made of grass, silages or legumes, is perfect for rumiants.",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Two spaces after period, please. Also, "rumiants" -> ruminants.

File: src/iuse.cpp [hunk 2]
```diff
@@ -1901,63 +1902,112 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
-        if(mon.type->id == (is_dogfood ? mon_dog : mon_cat)) {
-            p->add_msg_if_player(m_good, is_dogfood
-                ? _("The dog seems to like you!")
-                : _("The cat seems to like you!  Or maybe it just tolerates your presence better.  It's hard to tell with cats."));
-            mon.friendly = -1;
-            if( is_dogfood ) {
+        // This switch handles each petfood for each type of tameable monster.
+        switch( animal_food_type ) {
+        case DOGFOOD:
+
```
Comment (Night-Pryanik) [inline src/iuse.cpp]: Excess empty line.

File: src/iuse.cpp [hunk 3]
```diff
@@ -1901,63 +1902,112 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
+    // First a check to see if we are trying to feed a NPC dog food.
+    if( animal_food_type == DOGFOOD ) {
+        if( npc *const person_ = g->critter_at<npc>( dirp ) ) {
+            npc &person = *person_;
+            query_yn( _( "Are you sure you want to feed a person the dog food?" ) );
+            add_msg( _( "You put your %1$s into %2$s's mouth!" ), it->tname().c_str(), person.name.c_str() );
+
+            if( person.is_friend() || x_in_y( 9, 10 ) ) {
+                person.say(
+                    _( "Okay, but please, don't give me this again. I don't want to eat dog food in the cataclysm all day." ) );
```
Comment (Night-Pryanik) [inline src/iuse.cpp]: Two spaces after the period, please.

File: src/iuse.cpp [hunk 4]
```diff
@@ -1901,63 +1902,112 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
-        if( query_yn( is_dogfood ?
-            _( "Are you sure you want to feed a person the dog food?" ) :
-            _( "Are you sure you want to feed a person the cat food?" ) ) ) {
-            p->add_msg_if_player( _( "You put your %1$s into %2$s's mouth!" ), it->tname().c_str(), person.name.c_str() );
-            if( person.is_friend() || x_in_y( 9, 10 ) ) {
-                person.say( _( "Okay, but please, don't give me this again. I don't want to eat dog food in the cataclysm all day." ) );
+            break;
+        case CATTLEFODDER:
+            if( mon.has_flag( MF_MILKABLE )  ) {
+                add_msg( m_good, _( "The cow seems to like you! It lets you pat her and seems friendly." ) );
```
Comment (Night-Pryanik) [inline src/iuse.cpp]: Two spaces after the period, please.

Comment (AlecWhite) [inline src/iuse.cpp]: So it ends up as `mon.  has_flag( MF_MILKABLE )`?

File: src/iuse.cpp (hunk 3 above)
Comment (AlecWhite) [inline src/iuse.cpp]: Which period?

Comment (ZhilkinSerg) [inline src/iuse.cpp]: Period between sentences:

`... again. I don't...`

File: src/iuse.cpp [hunk 5]
```diff
@@ -1901,63 +1902,111 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
-        if( query_yn( is_dogfood ?
-            _( "Are you sure you want to feed a person the dog food?" ) :
-            _( "Are you sure you want to feed a person the cat food?" ) ) ) {
-            p->add_msg_if_player( _( "You put your %1$s into %2$s's mouth!" ), it->tname().c_str(), person.name.c_str() );
-            if( person.is_friend() || x_in_y( 9, 10 ) ) {
-                person.say( _( "Okay, but please, don't give me this again. I don't want to eat dog food in the cataclysm all day." ) );
+            break;
+        case CATTLEFODDER:
+            if( mon.has_flag( MF_MILKABLE ) ) {
+                add_msg( m_good, _( "The cow seems to like you! It lets you pat her and seems friendly." ) );
```
Comment (kevingranade) [inline src/iuse.cpp]: Switches between "it" and "her", which is awkward, either change "it" to "she" or rearrange the sentence to only have one pronoun.

File: src/iuse.cpp (hunk 4 above)
Comment (Night-Pryanik) [inline src/iuse.cpp]: Oh, sorry! I mean after the exclamation sign.

File: src/iuse.cpp [hunk 6]
```diff
@@ -1901,63 +1902,111 @@ int iuse::mycus(player *p, item *it, bool t, const tripoint &pos)
...
     }
     p->moves -= 15;
     const int mon_idx = g->mon_at( dirp, true );
-    if(mon_idx != -1) {
+
+    // First a check to see if we are trying to feed a NPC dog food.
+    if( animal_food_type == DOGFOOD ) {
+        if( npc *const person_ = g->critter_at<npc>( dirp ) ) {
+            npc &person = *person_;
+            query_yn( _( "Are you sure you want to feed a person the dog food?" ) );
```
Comment (BrettDong) [inline src/iuse.cpp]: If this query returns false, then the action should be canceled.

**Pull request merged**